keps/sig-scheduling/3521-pod-schedule-readiness/README.md

```diff
@@ -0,0 +1,987 @@
+-->
+
+This KEP aims to add a `.status.schedulePaused` field to Pod spec, to mark a Pod's schedule readiness.
+Integrators can mutate this field to instruct scheduler whether it's ready for scheduling or not.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Pods are naturally perceived ready for scheduling once created. Kubernetes scheduler does
+its due diligence to find nodes to place all pending Pods. However, in a real-world case,
+some Pods may stay in a "miss-essential-resources" state for a long period. These pods,
+actually churn the scheduler (and downstream integrators like Cluster AutoScaler) in an
+unnecessary manner.
+
+Lacking the knob to flag Pods as schedule-ready/unready wastes scheduling cycles on retrying
+Pods that are determined to be unschedulable. As a result, it delays the scheduling of other Pods,
+and also lowers the overall scheduling throughput. Moreover, it imposes restrictions to vendors to
+develop some in-house features (such as hierarchical quota) natively.
+
+This proposal comes up with an extensible mechanics to ensure Pods are scheduled only when they are
+semantically ready.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Define API to mark Pod as schedule-paused/unpaused.
+- Design a new Enqueue extension point to customize Pod's queueing behavior.
+- A default enqueue plugin to honor the new API semantics.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Update the Pod's condition if it carries `.status.schedulePaused=true`.
+- Focus on in-house use-cases of the Enqueue extension point.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+We propose a new field `.status.schedulePaused` to the Pod API. The field is defaulted to `false`.
+
+For Pods carrying `.status.schedulePause=true`, they will be "parked" in scheduler's internal
+unschedulablePods pool, and only gets retried when the field is set to `false`.
+
+Practically, this field can be initialized to `true` by MutatingWebhook, and be flipped to `false`
+by external integrators when certain criteria is met.
```

Comment:
If a user modifies this value post creation from false to true, but before spec.nodeName is set, should we reject spec.nodeName being set?

Put another way, is this a “one way” transition or not?  If not, how will schedulers leverage / react to that transition?

Reply:
> If a user modifies this value post creation from false to true, but before spec.nodeName is set, should we reject spec.nodeName being set?

Do you mean when the Pod is just **_being_** scheduled, and on the same time an in-flight update (change the value from false to true) comes in? (in other cases, the pod will be updated in-place inside scheduler queues)

> Put another way, is this a “one way” transition or not?

We don't restrict the mutation. Scheduler just looks at its actual state when it's about to be scheduled. And after the Pod gets scheduled (with nodeName set), scheduler doesn't care about that Pod anymore.

Reply:
> I think this should be a one-way transition from true to false. The field can be set to true only on create and never on update.

This would make our life easier - we don't need to worry about a being-scheduled Pod gets updated elsewhere with the paused flag from false to true.

My concern is: by restricting the transition, do we lose some flexibility? like, is the "reviving a pod" case mentioned in https://github.com/kubernetes/enhancements/pull/3522#discussion_r973441963 is a valid usecase? Want to hear more other real users. cc/ @wilfred-s

Reply:
If pod is waiting for scheduling, we can mutate the `spec.schedulePaused`, if pod is in scheduling, we disallow the transition.

Reply:
> If pod is waiting for scheduling, we can mutate the spec.schedulePaused, if pod is in scheduling, we disallow the transition.

It will be hard to be this granular because the scheduler and api-server don't communicate via shared memory, the state is communicated via the object itself and so it will be hard to avoid race conditions. I would start strict and relax later if the need arises.

Reply:
> if pod is in scheduling, we disallow the transition.

This needs to yield a signal to API Server from scheduler, and the only way to achieve this is through some API object's change. But I don't think it's practical to come up with another level of abstraction on the scheduling phase for a Pod.

Reply:
@ahg-g for unschedulable Pod, it makes sense to restrict the transition from true->false. But for scheduled Pod, let's brainstorm in https://github.com/kubernetes/enhancements/pull/3522#discussion_r974485890.

Reply:
Fold this thread as we almost reach a consensus in the thread of https://github.com/kubernetes/enhancements/pull/3522#discussion_r974485890.